tools/rewriter/DetermineAbi.cpp

```diff
@@ -204,32 +219,53 @@ cgFunctionInfo(clang::CodeGen::CodeGenModule &cgm,
   }
 }
 
+// Given a function signature, find the layout of the argument and return value slots.
 CAbiSignature determineAbi(const clang::CodeGen::CGFunctionInfo &info,
                            const clang::ASTContext &astContext) {
   // get ABI for return type and each parameter
   CAbiSignature sig;
   sig.variadic = info.isVariadic();
+
+  // We want to find the layout of the parameter and return value "slots."
+  // We can store a certain number of slots in registers, while the rest
+  // will go on the stack.
+  // These are "slots" and not parameters because some parameters larger
+  // than 64 bits can occupy multiple slots.
+  // Each slot is of one of the types described in CAbiArgKind.
+
+  // Get the slots for the return value.
   auto &returnInfo = info.getReturnInfo();
   sig.ret = abiSlotsForArg(info.getReturnType(), returnInfo, astContext);
 
   auto is_integral = [](auto &x) { return x == CAbiArgKind::Integral; };
+
+  // num_regs is the number of registers in which the return value is stored.
+  // This may be one for a value up to 64 bits or two for a 128-bit value.
   size_t num_regs = std::count_if(sig.ret.begin(), sig.ret.end(), is_integral);
-  if (num_regs > 2) {
+  // It's possible that clang gives us back a value of three or more integers,
+  // for example a struct full of ints. However, in reality the whole value
+  // goes on the stack in this case according to the x64 ABI.
+  // We handle that case here.
```

Comment:
I wish Clang documented which transformations it does and doesn't do by this point in compilation, but sadly we're really flying by the seat of our pants. I would maybe ask to qualify more what kind of "possible" this is--one where we don't know what Clang does right now (possibly for some types we haven't observed), or one where we don't know what Clang may do in the future.